homeassistant/components/wilight/switch.py

```diff
@@ -111,54 +111,48 @@ async def async_setup_entry(
     async_add_entities(entities)
 
     # Handle services for a discovered WiLight device.
-    async def set_watering_time(service: Any) -> None:
+    async def set_watering_time(entity, service: Any) -> None:
         entity_id = service.data[ATTR_ENTITY_ID]
 
-        for entity in entities:
-            if (entity.entity_id == entity_id) & isinstance(entity, WiLightValveSwitch):
-                watering_time = service.data[ATTR_WATERING_TIME]
-                await entity.async_set_switch_time(watering_time=watering_time)
+        if (entity.entity_id == entity_id) & isinstance(entity, WiLightValveSwitch):
+            watering_time = service.data[ATTR_WATERING_TIME]
+            await entity.async_set_watering_time(watering_time=watering_time)
 
-    async def set_trigger(service: Any) -> None:
+    async def set_trigger(entity, service: Any) -> None:
         entity_id = service.data[ATTR_ENTITY_ID]
 
-        for entity in entities:
-            if (entity.entity_id == entity_id) & isinstance(entity, WiLightValveSwitch):
-                trigger_index = service.data[ATTR_TRIGGER_INDEX]
-                trigger = service.data[ATTR_TRIGGER]
-                await entity.async_set_trigger(
-                    trigger_index=trigger_index, trigger=trigger
-                )
+        if (entity.entity_id == entity_id) & isinstance(entity, WiLightValveSwitch):
+            trigger_index = service.data[ATTR_TRIGGER_INDEX]
+            trigger = service.data[ATTR_TRIGGER]
+            await entity.async_set_trigger(trigger_index=trigger_index, trigger=trigger)
 
-    async def set_pause_time(service: Any) -> None:
+    async def set_pause_time(entity, service: Any) -> None:
         entity_id = service.data[ATTR_ENTITY_ID]
 
-        for entity in entities:
-            if (entity.entity_id == entity_id) & isinstance(
-                entity, WiLightValvePauseSwitch
-            ):
-                pause_time = service.data[ATTR_PAUSE_TIME]
-                await entity.async_set_switch_time(pause_time=pause_time)
+        if (entity.entity_id == entity_id) & isinstance(
+            entity, WiLightValvePauseSwitch
+        ):
+            pause_time = service.data[ATTR_PAUSE_TIME]
+            await entity.async_set_pause_time(pause_time=pause_time)
 
-    hass.services.async_register(
-        DOMAIN,
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
         SERVICE_SET_WATERING_TIME,
+        SERVICE_SCHEMA_WATERING_TIME,
```

Comment:
The service schema should just be a dict with voluptuous validators and not include a key for the entity_id. See the example here:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

Reply:
I tried removing the key for entity_id but couldn't get it to work without.
Removing:  vol.Required(ATTR_ENTITY_ID): cv.entity_id,
I get the error: ERROR (MainThread) [homeassistant.helpers.script.websocket_api_script] websocket_api script: Error executing script. Invalid data for call_service at pos 1: extra keys not allowed @ data['entity_id']
With vol.Required(ATTR_ENTITY_ID): cv.entity_id, the service works well.

Reply:
I found where I was going wrong.
OK, fixed.

Reply:
Done